lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp

```diff
@@ -305,6 +305,239 @@ class ConvertValsemVariantAtenIndexPutImplOp
+public:
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(AtenMaxPool2dWithIndicesBackwardOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    if (failed(verifyLinalgCompatibleTypes(op, rewriter)))
+      return failure();
+
+    Location loc = op.getLoc();
+    MLIRContext *context = op->getContext();
+    Value gradOutput = adaptor.grad_output();
+    Value input = adaptor.self();
+    RankedTensorType gradOutputType =
+        gradOutput.getType().cast<RankedTensorType>();
+    Type gradOutputElemType = gradOutputType.getElementType();
+    RankedTensorType inputType = input.getType().cast<RankedTensorType>();
+    Type inputElemType = inputType.getElementType();
+    int64_t tensorOperandRank = inputType.getRank();
+
+    // `TMTensor::ScatterOp` expects indices of element type i32.
+    Value indices = convertTensorToDtype(
+        rewriter, loc, op.indices(),
+        mlir::IntegerType::get(context, 32, mlir::IntegerType::Signed));
+    indices = typeConverter->materializeTargetConversion(
+        rewriter, loc, typeConverter->convertType(indices.getType()), indices);
+    RankedTensorType indicesType = indices.getType().cast<RankedTensorType>();
+    Type indicesElemType = indicesType.getElementType();
+
+    // The element type of the `input` and `values` should be same.
+    if (inputElemType != gradOutputElemType)
+      return rewriter.notifyMatchFailure(
+          op,
+          "Input element type should be same as the grad_output element type.");
+
+    // TODO: Add support for dynamic shaped operands.
+    if (!inputType.hasStaticShape() || !indicesType.hasStaticShape() ||
+        !gradOutputType.hasStaticShape())
+      return rewriter.notifyMatchFailure(
+          op, "grad_output, indices and input has to be static shaped.");
+
+    ArrayRef<int64_t> gradOutputShape(gradOutputType.getShape());
+    ArrayRef<int64_t> inputShape(inputType.getShape());
+    ArrayRef<int64_t> indicesShape(indicesType.getShape());
+
+    int64_t numelInput = 1, numelGradOutput = 1, numelIndices = 1;
+    for (unsigned i = 0; i < inputType.getRank(); i++) {
+      numelGradOutput *= gradOutputShape[i];
+      numelInput *= inputShape[i];
+      numelIndices *= indicesShape[i];
+    }
+
+    //  Since the scatter op requires indices to be a 2-d tensor, we first
+    // create a new 1-d tensor comprising the index values and then expand it
+    // from 1-d to a 2-d tensor. Also, the index values have to be updated to
+    // factor in the collapsed dimensions of the tensor. The algorithm for the
+    // updation of indices is as follows:
+    //
+    // for i in range(N):
+    //   for j in range(C):
+    //     for k in range(H):
+    //         for l in range(W):
+    //             updatedIndices[i * (C*H*W) + j * (H*W) + k * (W) + l] =
+    //             originalIndices[i, j, k, l] + j * (Hin*Win) + i * (C*Hin*Win)
+    //
```

Comment:
it might be easier just convert the indices to the form expected by the scatter op:
```
for i in range(N):
   for j in range(C):
      for k in range(H):
         for l in range(W):
            for m in range(4):
               m == 0:
                  updateIndices[i * (C*H*W) + j * (H*W) + k * (W) + l][0] = i
               m == 1:
                  updateIndices[i * (C*H*W) + j * (H*W) + k * (W) + l][1] = j
               m == 2:
                  updateIndices[i * (C*H*W) + j * (H*W) + k * (W) + l][2] = originalIndices[i, j, k, l] / Hin
               m == 3:
                  updateIndices[i * (C*H*W) + j * (H*W) + k * (W) + l][3] = originalIndices[i, j, k, l] % Hin
```
In this way the scatter op can be used directly without expanding the result in the end.